byzcoin/struct.go

```diff
@@ -1,12 +1,531 @@
+// oldest state change added (it will be the same if ordered).
+type stateChangeStorage struct {
+	db             *bolt.DB
+	bucket         []byte
+	size           int
+	maxSize        int
+	maxNbrBlock    int
+	sortedKeys     keyTimeArray
+	sortedKeysLock sync.Mutex
+}
+
+// Create a storage with a default maximum size
+func newStateChangeStorage(c *onet.Context) *stateChangeStorage {
+	db, name := c.GetAdditionalBucket(bucketStateChangeStorage)
+	return &stateChangeStorage{
+		db:      db,
+		bucket:  name,
+		maxSize: defaultMaxSize,
+	}
+}
+
+// init reads the db to recalculate the size and create a map
+// of the last known block indices for each skipchain
+func (s *stateChangeStorage) init() (map[string]int, error) {
+	s.size = 0
+	indices := make(map[string]int)
+
+	err := s.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket(s.bucket)
+		if b == nil {
+			return errors.New("Missing bucket")
+		}
+
+		// Each pair at this level is a bucket assigned to a skipchain
+		return b.ForEach(func(scid, v []byte) error {
+			scb := b.Bucket(scid)
+			if scb == nil {
+				return nil
+			}
+
+			var lastIndex int64
+
+			err := scb.ForEach(func(k, v []byte) error {
+				buf := bytes.NewBuffer(k[prefixLength+versionLength:])
+				var idx int64
+				err := binary.Read(buf, binary.BigEndian, &idx)
+				if err != nil {
+					return err
+				}
+
+				if idx > lastIndex {
+					lastIndex = idx
+				}
+
+				s.size += len(v)
+				return nil
+			})
+
+			indices[skipchain.SkipBlockID(scid).String()] = int(lastIndex + 1)
+			return err
+		})
+	})
+
+	return indices, err
+}
+
+// getBucket gets the bucket for the given skipchain
+func (s *stateChangeStorage) getBucket(tx *bolt.Tx, sid skipchain.SkipBlockID) *bolt.Bucket {
+	b := tx.Bucket(s.bucket)
+	if b == nil {
+		panic("Bucket has not been created. This is a programmer error.")
+	}
+
+	if tx.Writable() {
+		sbb, err := b.CreateBucketIfNotExists(sid)
+		if err != nil {
+			panic(err)
+		}
+
+		return sbb
+	}
+
+	return b.Bucket(sid)
+}
+
+// setMaxSize enables the cleaning of old state changes when the storage
+// size is above a given threshold. Note that the value is not strict.
+func (s *stateChangeStorage) setMaxSize(size int) {
+	s.maxSize = size
+}
+
+// setMaxNbrBlock enables the cleaning of state changes belonging
+// to blocks with an old index.
+func (s *stateChangeStorage) setMaxNbrBlock(nbr int) {
+	s.maxNbrBlock = nbr
+}
+
+// This will clean the oldest state changes when the total size
+// is above the maximum. It will remove elements until cleanThreshold of
+// the space is available.
+func (s *stateChangeStorage) cleanBySize() error {
+	if s.size < s.maxSize || s.maxSize == 0 {
+		// nothing to clean
+		return nil
+	}
+
+	err := s.db.Update(func(tx *bolt.Tx) error {
+		// We make enough space to not have to do it everytime
+		// we append state changes
+		thres := int(float64(s.maxSize) * cleanThreshold)
+
+		for s.size > thres {
+			if len(s.sortedKeys) == 0 {
```

Comment:
missing a mutex?

Reply:
No because the function is used inside the append that is already holding the lock. But I gonna change that for a read/write lock.

Reply:
No but I'm going to change that so that if you use the cleaning elsewhere it will hold the lock anyway, that will be less confusing.